src/main/java/io/wisetime/connector/fetch_client/FetchClient.java

```diff
@@ -0,0 +1,131 @@
+/*
+ * Copyright (c) 2019 Practice Insight Pty Ltd. All Rights Reserved.
+ */
+
+package io.wisetime.connector.fetch_client;
+
+import org.joda.time.DateTime;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.atomic.AtomicReference;
+
+import io.wisetime.connector.TimePosterRunner;
+import io.wisetime.connector.api_client.ApiClient;
+import io.wisetime.connector.api_client.PostResult;
+import io.wisetime.connector.integrate.WiseTimeConnector;
+import io.wisetime.generated.connect.TimeGroupStatus;
+
+/**
+ * Implements a fetch based approach to retrieve time groups.
+ *
+ * @author pascal.filippi@staff.wisetime.com
+ */
+public class FetchClient implements Runnable, TimePosterRunner {
+
+  private static final int MAX_MINS_SINCE_SUCCESS = 10;
+
+  private static final Logger log = LoggerFactory.getLogger(FetchClient.class);
+
+  private ApiClient apiClient;
+
+  private TimeGroupIdStore timeGroupIdStore;
+
+  private String fetchClientId;
+
+  private int limit;
+
+  private WiseTimeConnector connector;
+
+  private ExecutorService postTimeExecutor;
+
+  private final AtomicReference<DateTime> lastSuccessfulRun;
+
+  private ExecutorService fetchClientExecutor;
+
+  public FetchClient(ApiClient apiClient,
+                     WiseTimeConnector connector,
+                     TimeGroupIdStore timeGroupIdStore,
+                     String fetchClientId,
+                     int limit) {
+    this.apiClient = apiClient;
+    this.connector = connector;
+    this.timeGroupIdStore = timeGroupIdStore;
+    this.fetchClientId = fetchClientId;
+    this.limit = limit;
+    this.postTimeExecutor = Executors.newFixedThreadPool(limit);
+    this.lastSuccessfulRun = new AtomicReference<>(DateTime.now());
+  }
+
+  @Override
+  public void run() {
+    while (!Thread.currentThread().isInterrupted()) {
+      try {
+        apiClient.fetchTimeGroups(fetchClientId, limit).forEach(timeGroup ->
+            postTimeExecutor.submit(() -> {
+              if (!timeGroupIdStore.alreadySeen(timeGroup.getGroupId())) {
+                timeGroupIdStore.putTimeGroupId(timeGroup.getGroupId());
+                PostResult result = connector.postTime(null, timeGroup);
+                try {
+                  // TRANSIENT_FAILURE doesn't need to be handled.
+                  // simply sending no update is considered a transient failure after a certain timeout
+                  switch (result) {
+                    case SUCCESS:
+                      apiClient.updatePostedTimeStatus(new TimeGroupStatus()
+                          .status(TimeGroupStatus.StatusEnum.SUCCESS)
+                          .timeGroupId(timeGroup.getGroupId())
+                          .fetchClientId(fetchClientId));
```

Comment:
The GUI app should show what happens behind the scenes, so we need meaningful `log.info` for any of the result cases. Also there are events that should be triggered on success, please check how they are used in other places - `TAGS_UPSERTED` and `TIME_GROUP_POSTED.

Reply:
Thanks, will add it